File: timus/1152/main.cpp
#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <iostream>
#include <sstream>
#include <vector>

class TResult {
public:
    TResult()
        : MinimalDamage(UINT64_MAX) {
    }

    void MaybeUpdate(std::size_t CurrentDamage) {
        MinimalDamage = std::min(MinimalDamage, CurrentDamage);
    }

    bool PossibleImrovement(std::size_t CurrentDamage) const {
        return CurrentDamage < MinimalDamage;
    }

    std::size_t GetValue() const {
        return MinimalDamage;
    }

private:
    std::size_t MinimalDamage;
};

struct TBalcony {
    std::size_t Monsters;
    bool Destroyed;
};

class TDisposition {
public:
    TDisposition(std::vector<TBalcony> balconies)
        : Balconies(std::move(balconies))
        , CumulativeDamage(0){};

private:
    std::size_t CumulativeDamage;
    std::vector<TBalcony> Balconies;

    std::string Dump() const {
        std::stringstream ss;
        ss << "[";
        for (const auto& b : Balconies) {
            if (b.Destroyed) {
                ss << " - ";
            } else {
                ss << " " << b.Monsters << " ";
            }
        }
        ss << "]";
        return ss.str();
    }

    void GetNonDestroyedBalconyIxs(std::vector<std::size_t>& ixs) const {
        ixs.clear();
        ixs.reserve(Balconies.size());
        for (std::size_t ix = 0; ix < Balconies.size(); ix++) {
            if (!Balconies[ix].Destroyed) {
                ixs.push_back(ix);
            }
        }
    }

    TDisposition Shoot(std::size_t start) const {
        auto newDisposition = *this;

        for (std::size_t i = 0; i < 3; i++) {
            newDisposition.Balconies[(start + i) % Balconies.size()].Destroyed = true;
        }

        for (const auto& b : newDisposition.Balconies) {
            if (!b.Destroyed) {
                newDisposition.CumulativeDamage += b.Monsters;
            }
        }

        return newDisposition;
    }

public:
    void Step(TResult& result) {
        // std::cout << "STEP A" << Dump() << " " << CumulativeDamage << std::endl;

        std::vector<std::size_t> starting_points;
        GetNonDestroyedBalconyIxs(starting_points);

        if (starting_points.size() == 0) {
            result.MaybeUpdate(CumulativeDamage);
            return;
        }

        for (auto start : starting_points) {
            TDisposition newDisposition = Shoot(start);
            // std::cout << "STEP B " << start << " " << newDisposition.Dump() << " " << newDisposition.CumulativeDamage
            //           << std::endl;
            if (result.PossibleImrovement(CumulativeDamage)) {
                // std::cout << "STEP C" << std::endl;
                newDisposition.Step(result);
            }
        }
    }
};

int main() {
    std::size_t N;
    std::cin >> N;

    std::vector<TBalcony> balconies;
    std::size_t monsters;
    for (; N > 0; N--) {
        std::cin >> monsters;
        balconies.push_back(TBalcony{.Monsters = monsters, .Destroyed = false});
    }

    TResult result;

    TDisposition(balconies).Step(result);

    std::cout << result.GetValue() << std::endl;
}

File: timus/1152/main.py
#!/usr/bin/env python3

import sys
from dataclasses import dataclass
from typing import List, Set
import copy
import fileinput


@dataclass
class Result:
    min_damage: int

    solution_cache: Set[str]

    def maybe_update(self, damage: int):
        self.min_damage = min(self.min_damage, damage)


@dataclass
class Balcony:
    monsters: int
    destroyed: bool


@dataclass
class Disposition:
    cum_damage: int
    balconies: List[Balcony]

    def __str__(self) -> str:
        return self.__dump()

    def __dump(self) -> str:
        s = "["
        for b in self.balconies:
            if b.destroyed:
                s += " - "
            else:
                s += f" {b.monsters} "
        s += "]"
        return s

    def non_destroyed_balcony_ixs(self) -> List[int]:
        return [
            ix for ix in range(len(self.balconies)) if not self.balconies[ix].destroyed
        ]

    def step(self, result: Result):
        starting_points = self.non_destroyed_balcony_ixs()

        if len(starting_points) == 0:
            result.maybe_update(self.cum_damage)
            return

        for start in starting_points:
            disp = self.shoot(start=start)
            # print(disp.cum_damage, result.min_damage)
            if disp.cum_damage < result.min_damage:
                disp.step(result=result)

    def shoot(self, start: int) -> "Disposition":
        balconies = copy.deepcopy(self.balconies)

        i = 0
        while i < 3:
            balconies[self._ix(start + i)].destroyed = True
            i += 1

        added_damage = 0
        for balcony in balconies:
            if not balcony.destroyed:
                added_damage += balcony.monsters

        return Disposition(
            balconies=balconies, cum_damage=self.cum_damage + added_damage
        )

    def _ix(self, i: int) -> int:
        return i % len(self.balconies)


if __name__ == "__main__":
    lines = list(fileinput.input())
    N = int(lines[0])
    monsters = list(map(lambda x: int(x), lines[1].split()))
    balconies = [Balcony(monsters=m, destroyed=False) for m in monsters]
    disp = Disposition(cum_damage=0, balconies=balconies)
    result = Result(min_damage=sys.maxsize, solution_cache=set())
    disp.step(result=result)
    print(result.min_damage)
